### PR TITLE
cucumber-expressions - TreeRegexp: recognize lookaround as non-capturing groups

### PR DESCRIPTION
**Pull request by vincent-psarga**
## Summary

This should provide a solution for  issue #643

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue).
- [ ] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).

## Checklist:

- [x] The change has been ported to Java.
- [x] The change has been ported to Ruby.
- [x] The change has been ported to JavaScript.
- [x] The change has been ported to Go.
- [ ] The change has been ported to .NET.
- [x] I've added tests for my code.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by aslakhellesoy** [comment]
> Can we try keep the tests standardised

We already have `/cucumber-expressions/examples.txt` for shared tests, and I guess we should use that more. If you edit that file, sync it to the sub folders:

```
source ../../scripts/functions.sh && rsync_files
```
WDYT @luke-hill ?

**Comment by vincent-psarga** [comment]
Just to ensure I understood @luke-hill comment correctly, on the Ruby example, you'd prefer me to rewrite:

```ruby
      it 'recognizes lookaround capture groups' do
        tr = TreeRegexp.new(/foo is a (?=[bar])(.*)$/)
        group = tr.match("foo is a big bar")
        expect(group.children[0].value).to eq('big bar')
      end
```

as 

```ruby
      it 'ignores `?=` as a non-capturing group' do
        tr = TreeRegexp.new(/a(?!b)(.+)/)
        group = tr.match('aBc')
        expect(group.value).to eq('aBc')
        expect(group.children[0].value).to eq('Bc')
        expect(group.children.length).to eq 1
      end
```

Is that right ? (and also for Java of course)

For the examples in examples.txt, I can add that:

```
/I have cukes in my (?=[b])(.*)$/
I have cukes in my belly
["belly"]
---
/I have cukes in my (?=[b])(.*)$/
I have cukes in my stomach
null
```

Although it is not 100% clear when reading that what we exactly try to match there, I like the idea of having clear tests describing what we are checking.
Having both of them makes more sense I think.

**Comment by vincent-psarga** [comment]
Note about examples.txt: it may not be possible to add the examples there, as Go does not support some examples we want to exhibit here (typically ```(?=re)```, but maybe the other ones that were described in #643)

See: https://github.com/google/re2/wiki/Syntax

**Comment by luke-hill** [comment]
Sounds like a plan RE standardised tests. But I've no idea how to implement them. I'm still reeling from the mass of work I had to do to get the last ones passing, so if it's easy to implement I'm for it.

cc/ @aslakhellesoy

**Comment by luke-hill** [comment]
Also RE the re-writing of tests. Can we have everything inside one capture, not 2. As my understanding of look aheads e.t.c. is that they don't capture inside their group. So if you have a blank group it kinda negates it.

**Comment by vincent-psarga** [comment]
> Can we have everything inside one capture, not 2

That sound doable :)

I'll try to complete the support of ```?<=``` before marking as ready :)

**Comment by vincent-psarga** [comment]
I've added support for ```?<=``` and ```?<!```in Ruby.

That said I'm not 100% convinced by the way it's done (seems pretty brittle to me). If anyone as a better idea, it's much welcome :)

(I though about looking the last two characters but I get the exception about named groups getting raised before matching ```=``` or ```!```)

**Comment by vincent-psarga** [comment]
@luke-hill is your change request still relevant after the changes I did or did I miss something else ?

File: cucumber-expressions/ruby/lib/cucumber/cucumber_expressions/tree_regexp.rb [hunk 1]
```diff
@@ -55,6 +59,18 @@ def match(s)
         group_indices = (0..match.length).to_a.to_enum
         @group_builder.build(match, group_indices)
       end
+
+      private
+
+      def end_non_capturing_group
+        @stack.last.set_non_capturing!
+        end_group()
```
Comment (luke-hill) [inline cucumber-expressions/ruby/lib/cucumber/cucumber_expressions/tree_regexp.rb]: here and in a few others places we have redundant `()`

**Comment by vincent-psarga** [comment]
Go does not support look around so that's why it's not checked here (well, the Go regex parser will fail when seeing lookarounds)
And for .NET, I guess the expressions are in Specflow, not this repo.

The redundant parenthesis should be removed now :)

**Comment by aslakhellesoy** [comment]
> And for .NET, I guess the expressions are in Specflow, not this repo.

There is no cucumber-expressions implementation in .NET yet

**Comment by luke-hill** [comment]
Genuine question: do we need something like this: https://github.com/cucumber/cucumber/pull/593/files#diff-992ea8ba4a1d5037dd6de00fe673baddR28

If not this is RtM (CI passing obviously)

**Comment by mpkorstanje** [comment]
> Genuine question: do we need something like this: https://github.com/cucumber/cucumber/pull/593/files#diff-992ea8ba4a1d5037dd6de00fe673baddR28

Not really. Checks that don't try to create the regex are prone to false positives. I've removed it.

**Comment by mpkorstanje** [comment]
I think we're good to go. 

If nobody has any remarks I'll merge this by the end of the week. But please feel free to do so earlier if you're think we're good to go.

**Comment by aslakhellesoy** [comment]
> @aslakhellesoy looks like something external to this MR changed and broke gherkin-god and the dots-formatter

Yes, a time-bomb caused the build to break a week or so ago. I've fixed that on master, and merging master into this one should fix it.

**Pull request merged**